File: tools/realboy_compare.go
package main

import (
	"encoding/hex"
	"fmt"
	"log"
	"os"
	"os/exec"
	"reflect"
	"regexp"
	"strings"
	"syscall"
	"time"
)

type Regs struct {
	AF uint16
	BC uint16
	DE uint16
	HL uint16
	SP uint16
	PC uint16
}

func (r *Regs) Set(name string, value uint16) {
	switch name {
	case "AF":
		r.AF = value
	case "BC":
		r.BC = value
	case "DE":
		r.DE = value
	case "HL":
		r.HL = value
	case "SP":
		r.SP = value
	case "PC":
		r.PC = value
	}
}

func (r *Regs) Flags() string {
	var flags []string
	if r.AF&0x80 != 0 {
		flags = append(flags, "Z")
	}
	if r.AF&0x40 != 0 {
		flags = append(flags, "N")
	}
	if r.AF&0x20 != 0 {
		flags = append(flags, "H")
	}
	if r.AF&0x10 != 0 {
		flags = append(flags, "C")
	}
	return strings.Join(flags, "")
}

func main() {
	p1 := exec.Command("target/debug/gameboy-emu", "-b", "res/DMG_ROM.bin", "res/cpu_instrs/cpu_instrs.gb", "--debug")
	p1.SysProcAttr = &syscall.SysProcAttr{Pdeathsig: syscall.SIGKILL}

	stdin1, err := p1.StdinPipe()
	if err != nil {
		log.Fatal(err)
	}
	defer stdin1.Close()

	stdout1, err := p1.StdoutPipe()
	if err != nil {
		log.Fatal(err)
	}
	defer stdout1.Close()

	p1.Stderr = os.Stderr

	p2 := exec.Command("/home/outroot/build/realboy-0.2.2/src/realboy", "-d", "res/cpu_instrs/cpu_instrs.gb")
	p2.SysProcAttr = &syscall.SysProcAttr{Pdeathsig: syscall.SIGKILL}

	stdin2, err := p2.StdinPipe()
	if err != nil {
		log.Fatal(err)
	}
	defer stdin2.Close()

	stdout2, err := p2.StdoutPipe()
	if err != nil {
		log.Fatal(err)
	}
	defer stdout2.Close()

	p2.Stderr = os.Stderr

	if err = p1.Start(); err != nil {
		log.Fatal(err)
	}

	if err = p2.Start(); err != nil {
		log.Fatal(err)
	}

	defer func() {
		p1.Process.Signal(os.Kill)
		p2.Process.Signal(os.Kill)
	}()

	time.Sleep(1 * time.Second)
	stdin1.Write([]byte("b 81d\n"))
	stdin1.Write([]byte("c\n"))
	stdin2.Write([]byte("break 0x81d\n"))
	stdin2.Write([]byte("step 0xFFFFFFFF\n"))
	time.Sleep(3 * time.Second)

	buf1 := make([]byte, 2*1024)
	buf2 := make([]byte, 2*1024)

	cont := make(chan bool)
	go func() {
		for {
			nr1, err := stdout1.Read(buf1)
			if err != nil {
				log.Fatal(err)
			}

			if nr1 > 0 {
				cont <- true
				break
			}
		}

		for {
			nr2, err := stdout2.Read(buf2)
			if err != nil {
				log.Fatal(err)
			}

			if nr2 > 0 {
				cont <- true
				break
			}
		}
	}()

	<-cont
	<-cont

	rx1 := regexp.MustCompile(`(?m:^([AFBCDEHLSPC]{2}):\s+0x(....).*$)`)
	rx2 := regexp.MustCompile(`(?m:^([AFBCDEHLSPC]{2}) = 0x(....).*$)`)
	go func() {
		for {
			stdin1.Write([]byte("debug\n"))
			stdin2.Write([]byte("show regs\n"))

			time.Sleep(100 * time.Millisecond)

			nr1, err := stdout1.Read(buf1)
			if err != nil {
				log.Fatal(err)
			}

			reg1 := Regs{}
			for _, v := range rx1.FindAllSubmatch(buf1[:nr1], -1) {
				tmp := make([]byte, 2)
				hex.Decode(tmp, v[2])

				reg1.Set(string(v[1]), uint16(tmp[0])<<8|uint16(tmp[1]))
			}

			nr2, err := stdout2.Read(buf2)
			if err != nil {
				log.Fatal(err)
			}

			reg2 := Regs{}
			for _, v := range rx2.FindAllSubmatch(buf2[:nr2], -1) {
				tmp := make([]byte, 2)
				hex.Decode(tmp, v[2])

				reg2.Set(string(v[1]), uint16(tmp[0])<<8|uint16(tmp[1]))
			}

			if !reflect.DeepEqual(reg1, reg2) {
				fmt.Printf("AF: Got: %04x, Expect: %04x\n", reg1.AF, reg2.AF)
				fmt.Printf("BC: Got: %04x, Expect: %04x\n", reg1.BC, reg2.BC)
				fmt.Printf("DE: Got: %04x, Expect: %04x\n", reg1.DE, reg2.DE)
				fmt.Printf("HL: Got: %04x, Expect: %04x\n", reg1.HL, reg2.HL)
				fmt.Printf("SP: Got: %04x, Expect: %04x\n", reg1.SP, reg2.SP)
				fmt.Printf("PC: Got: %04x, Expect: %04x\n", reg1.PC, reg2.PC)
				fmt.Printf("Flags: Got: %s, Expect: %s\n", reg1.Flags(), reg2.Flags())
				os.Exit(0)
			}

			fmt.Printf("%04x\n", reg1.PC)
			stdin1.Write([]byte("s\n"))
			stdin2.Write([]byte("step\n"))
		}
	}()

	p1.Wait()
	p2.Wait()
}

File: tools/realboy_compare.py
#!python3.5

import pexpect
import re
import sys
import traceback

def flags(value):
  flags = []
  if value&0x80 != 0:
  	flags.append("Z")
  if value&0x40 != 0:
  	flags.append("N")
  if value&0x20 != 0:
  	flags.append("H")
  if value&0x10 != 0:
  	flags.append("C")
  return "".join(flags)

child1 = pexpect.spawn('target/debug/gameboy-emu --debug "res/cpu_instrs/cpu_instrs.gb"')
child2 = pexpect.spawn('/home/outroot/build/realboy-0.2.2/src/realboy -d "res/cpu_instrs/cpu_instrs.gb"')

try:
  # class logger:
  #   def write(data):
  #     print(data.decode())
  #   def flush():
  #     sys.stdout.flush()
  #
  # child1.logfile = logger
  # child2.logfile = logger

  child1.expect('(gameboy)')
  child1.sendline('b 101') # c920 c0d2 c7f3 c8a7 c8db, cc41
  child1.expect('(gameboy)')
  child1.sendline('c')
  child1.expect('Breakpoint hit.*?: ')
  child1.expect('(gameboy)')

  inst1 = child1.before.decode().split('\r')[0]

  child2.expect('gddb>')
  child2.sendline('break 0x101')
  child2.expect('gddb>')
  child2.sendline('step 0xFFFFFFFF')
  child2.expect('Breakpoint')
  child2.expect('0x')
  child2.expect('gddb>')

  split2 = child2.before.decode().split('\t')
  pc = "0x" + split2[0].split('\r')[0]
  inst2 = split2[1]

  while True:
    print("{}: {}  <->  {}".format(pc, inst1, inst2))

    child1.sendline('debug')
    child1.expect('(gameboy)')
    regex1 = re.compile('([AFBCDEHLSPC]{2}):\s+0x(....).*?', re.MULTILINE)
    match1 = list(regex1.findall(child1.before.decode()))

    child2.sendline('show regs')
    child2.expect('gddb>')
    regex2 = re.compile('([AFBCDEHLSPC]{2}) = 0x(....).*?', re.MULTILINE)
    match2 = list(regex2.findall(child2.before.decode()))

    if pc == '0xc7f5' or pc == '0x745':
      match1[0] = match2[0]
      child1.sendline('set af {}'.format(match2[0][1]))
      child1.expect('(gameboy)')

    fail = False
    for i in range(len(match1)):
      if match1[i][1] != match2[i][1]:
        # if i == 0 and match1[i][1][2:] == match2[i][1][2:]:
        #   continue
        # if i == 0:
        #   continue
        fail = True
        break

    if fail:
      print("AF: Got: {}, Expect: {}".format(match1[0][1], match2[0][1]))
      print("BC: Got: {}, Expect: {}".format(match1[1][1], match2[1][1]))
      print("DE: Got: {}, Expect: {}".format(match1[2][1], match2[2][1]))
      print("HL: Got: {}, Expect: {}".format(match1[3][1], match2[3][1]))
      print("SP: Got: {}, Expect: {}".format(match1[4][1], match2[4][1]))
      print("PC: Got: {}, Expect: {}".format(match1[5][1], match2[5][1]))
      f1 = int(match1[0][1], 16)
      f2 = int(match2[0][1], 16)
      print("Flags: Got: {}, Expect: {}\n\n".format(flags(f1), flags(f2)))
      sys.exit(1)

    child1.sendline('s')
    child1.expect('(gameboy)')
    split1 = re.split('[\r\n]+', child1.before.decode())
    split1 = split1[2].split(': ')
    pc = split1[0]
    inst1 = split1[1]

    child2.sendline('step')
    child2.expect('gddb>')
    split2 = child2.before.decode().split('\r\n')
    split2 = split2[1].split('\t')
    inst2 = split2[1]

except BaseException as error:
  print('An exception occurred: {}'.format(error))
  print(traceback.format_exc())
  # print(str(child1))
  # print(str(child2))
